[ML] Switch from normal sampling to random sampler for Index data visualizer table

qn895 opened the pull request:
## Summary

This PR switches the currently sampling method from normal sampling to random sampler for Index data visualizer table. It also lowers the threshold for when the sampling can be done in the interest of speed.

<img width="1777" alt="Screen Shot 2022-11-04 at 16 01 31" src="https://user-images.githubusercontent.com/43350163/200074204-92f8d314-076d-4ab9-be5c-06d58f813639.png">


### Checklist

Delete any items that are not applicable to this PR.

- [ ] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [ ] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [ ] Any UI touched in this PR is usable by keyboard only (learn more about [keyboard accessibility](https://webaim.org/techniques/keyboard/))
- [ ] Any UI touched in this PR does not create any new axe failures (run axe in browser: [FF](https://addons.mozilla.org/en-US/firefox/addon/axe-devtools/), [Chrome](https://chrome.google.com/webstore/detail/axe-web-accessibility-tes/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US))
- [ ] If a plugin configuration key changed, check if it needs to be allowlisted in the cloud and added to the [docker list](https://github.com/elastic/kibana/blob/main/src/dev/build/tasks/os_packages/docker_generator/resources/base/bin/kibana-docker)
- [ ] This renders correctly on smaller devices using a responsive layout. (You can test this [in your browser](https://www.browserstack.com/guide/responsive-testing-on-local-server))
- [ ] This was checked for [cross-browser compatibility](https://www.elastic.co/support/matrix#matrix_browsers)


### Risk Matrix

Delete this section if it is not applicable to this PR.

Before closing this PR, invite QA, stakeholders, and other developers to identify risks that should be tested prior to the change/feature release.

When forming the risk matrix, consider some of the following examples and how they may potentially impact the change:

| Risk                      | Probability | Severity | Mitigation/Notes        |
|---------------------------|-------------|----------|-------------------------|
| Multiple Spaces&mdash;unexpected behavior in non-default Kibana Space. | Low | High | Integration tests will verify that all features are still supported in non-default Kibana Space and when user switches between spaces. |
| Multiple nodes&mdash;Elasticsearch polling might have race conditions when multiple Kibana nodes are polling for the same tasks. | High | Low | Tasks are idempotent, so executing them multiple times will not result in logical error, but will degrade performance. To test for this case we add plenty of unit tests around this logic and document manual testing procedure. |
| Code should gracefully handle cases when feature X or plugin Y are disabled. | Medium | High | Unit tests will verify that any feature flag or plugin combination still results in our service operational. |
| [See more potential risk examples](https://github.com/elastic/kibana/blob/main/RISK_MATRIX.mdx) |


### For maintainers

- [ ] This was checked for breaking API changes and was [labeled appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)

elasticmachine commented:
Pinging @elastic/ml-ui (:ml)

benwtrent commented on x-pack/plugins/data_visualizer/public/application/index_data_visualizer/search_strategy/requests/get_document_stats.ts:
```diff
@@ -189,13 +187,9 @@ export const getDocumentCountStats = async (
-      newProbability === Infinity ||
-      numSampled / initialDefaultProbability < 1e7
-    ) {
+    if (numSampled === 0 || newProbability === Infinity || numSampled < 5) {
```
Why is `5` chosen here? It seems to me it should be much bigger. 

Also, `numSampled === 0` already implies `numSampled < 5`, so that equals `0` check is redundant.

benwtrent commented on x-pack/plugins/data_visualizer/public/application/index_data_visualizer/search_strategy/requests/get_numeric_field_stats.ts:
```diff
@@ -158,24 +147,17 @@ export const fetchNumericFieldsStats = (
             min: get(fieldStatsResp, 'min', 0),
             max: get(fieldStatsResp, 'max', 0),
             avg: get(fieldStatsResp, 'avg', 0),
             isTopValuesSampled:
               field.cardinality >= SAMPLER_TOP_TERMS_THRESHOLD || samplerShardSize > 0,
```
This logic seems wrong now? I don't think you use `SAMPLER_TOP_TERMS_THRESHOLD` anylonger when determining if a thing is sampled or not.

benwtrent reviewed (commented):
Are there any safety checks for small data?

walterra commented:
For total documents we recalculate now an estimated number based on the sampling. Do we aim for doing the same for all other data too in the table so that for example the documents stats would also be estimated full numbers recalculated from the sampling probability?

peteharverson commented:
> For total documents we recalculate now an estimated number based on the sampling. Do we aim for doing the same for all other data too in the table so that for example the documents stats would also be estimated full numbers recalculated from the sampling probability?

Currently the counts in the expanded rows are not consistent:

<img width="719" alt="image" src="https://user-images.githubusercontent.com/7405507/201731822-0ef1faf6-4c7e-4b24-bbc6-8b414293e56d.png">

The doc stats add up to the total doc count, whereas the top values add up to the sampled count. We should be consistent here - either using the total count or the sampled count in both places. The previous approached used the sampled total.

<img width="930" alt="image" src="https://user-images.githubusercontent.com/7405507/201732143-8ebb9d77-f60f-4fd3-ad6e-929b87da8ef4.png">

jgowdyelastic commented on x-pack/plugins/data_visualizer/public/application/common/components/top_values/top_values.tsx:
```diff
@@ -53,70 +52,48 @@ export const TopValues: FC<Props> = ({ stats, fieldFormat, barColor, compressed,
+  const totalDocuments = stats.totalDocuments ?? 0;
+  const topValuesOtherCountPercent =
+    1 - (topValues ? topValues.reduce((acc, bucket) => acc + bucket.percent, 0) : 0);
+  const topValuesOtherCount = Math.floor(topValuesOtherCountPercent * (sampleCount || 0));
```
although it doesn't matter logically here, for consistency it might be better to still use `??` because the point of the check is to look for `undefined`

qn895 replied:
Fixed here [`80872e0` (#144646)](https://github.com/elastic/kibana/pull/144646/commits/80872e03e0de597573a005988aa03417e83cfb5a)

jgowdyelastic commented on x-pack/plugins/data_visualizer/public/application/index_data_visualizer/search_strategy/requests/build_random_sampler_agg.ts:
```diff
@@ -0,0 +1,120 @@
+export type FieldStatsEmbeddableSamplerOption =
+  typeof EMBEDDABLE_SAMPLER_OPTION[keyof typeof EMBEDDABLE_SAMPLER_OPTION];
+
+export function isRandomSamplingOption(arg: SamplingOption): arg is RandomSamplingOption {
```
these guards might be better living next to the types in `common/types/field_stats`

qn895 replied:
Fixed here [`80872e0` (#144646)](https://github.com/elastic/kibana/pull/144646/commits/80872e03e0de597573a005988aa03417e83cfb5a)

jgowdyelastic commented on x-pack/plugins/data_visualizer/public/application/index_data_visualizer/search_strategy/requests/build_random_sampler_agg.ts:
```diff
@@ -0,0 +1,120 @@
+}
+
+export function buildAggregationWithSamplingOption(
+  aggs: any,
```
can this `any` be replaced by a correct type?
maybe `Record<string, estypes.AggregationsAggregationContainer>`

qn895 replied:
Fixed here [`80872e0` (#144646)](https://github.com/elastic/kibana/pull/144646/commits/80872e03e0de597573a005988aa03417e83cfb5a)

jgowdyelastic commented on x-pack/plugins/data_visualizer/public/application/index_data_visualizer/search_strategy/requests/build_random_sampler_agg.ts:
```diff
@@ -0,0 +1,120 @@
+}
+
+export function buildSamplerAggregation(
+  aggs: any,
```
same with this `any`

qn895 replied:
Fixed here https://github.com/elastic/kibana/pull/144646/commits/80872e03e0de597573a005988aa03417e83cfb5a

jgowdyelastic commented on x-pack/plugins/data_visualizer/public/application/index_data_visualizer/search_strategy/requests/build_random_sampler_agg.ts:
```diff
@@ -0,0 +1,120 @@
+ * Wraps the supplied aggregations in a random sampler aggregation.
+ */
+export function buildRandomSamplerAggregation(
+  aggs: any,
```
same with this `any`

qn895 replied:
Fixed here https://github.com/elastic/kibana/pull/144646/commits/80872e03e0de597573a005988aa03417e83cfb5a

jgowdyelastic commented:
It would be good to add a debounce or delay to the sampling percentage slider.
It goes a bit crazy when moving it.

![635FA12B-7B4E-4B84-B0FB-641E5A072014-39619-0001669D073D610D](https://user-images.githubusercontent.com/22172091/201748715-14e7cbaa-cebc-49e9-b02b-91612fedd3cb.gif)

qn895 commented on x-pack/plugins/data_visualizer/public/application/common/components/top_values/top_values.tsx:
```diff
@@ -36,8 +36,7 @@ interface Props {
 
-function getPercentLabel(docCount: number, topValuesSampleSize: number): string {
-  const percent = (100 * docCount) / topValuesSampleSize;
+function _getPercentLabel(percent: number): string {
```
Fixed here [`80872e0` (#144646)](https://github.com/elastic/kibana/pull/144646/commits/80872e03e0de597573a005988aa03417e83cfb5a)

qn895 commented on x-pack/plugins/data_visualizer/public/application/common/components/stats_table/components/field_data_row/document_stats.tsx:
```diff
@@ -8,32 +8,46 @@
+export const DocumentStat = ({ config, showIcon, totalCount }: Props) => {
   const { stats } = config;
+  const {
+    services: { data },
```
Fixed here [`80872e0` (#144646)](https://github.com/elastic/kibana/pull/144646/commits/80872e03e0de597573a005988aa03417e83cfb5a)

qn895 commented on x-pack/plugins/data_visualizer/public/application/common/components/stats_table/components/field_data_expanded_row/document_stats.tsx:
```diff
@@ -46,6 +46,13 @@ export const DocumentStatsTable: FC<FieldDataRowProps> = ({ config }) => {
+  const docsPercent =
+    valueCount !== undefined && sampleCount !== undefined
+      ? roundToDecimalPlace((valueCount / sampleCount) * 100)
+      : 0;
```
Fixed here [`80872e0` (#144646)](https://github.com/elastic/kibana/pull/144646/commits/80872e03e0de597573a005988aa03417e83cfb5a)

qn895 commented on x-pack/plugins/data_visualizer/public/application/common/components/stats_table/components/field_data_expanded_row/choropleth_map.tsx:
```diff
@@ -97,13 +99,55 @@ interface Props {
 export const ChoroplethMap: FC<Props> = ({ stats, suggestion }) => {
-  const { fieldName, isTopValuesSampled, topValues, topValuesSamplerShardSize } = stats!;
+  const {
+    services: { data },
```
Fixed here [`80872e0` (#144646)](https://github.com/elastic/kibana/pull/144646/commits/80872e03e0de597573a005988aa03417e83cfb5a)

qn895 commented on x-pack/plugins/data_visualizer/public/application/common/components/stats_table/components/field_data_row/document_stats.tsx:
```diff
@@ -8,32 +8,46 @@
   // If field exists is docs but we don't have count stats then don't show
   // Otherwise if field doesn't appear in docs at all, show 0%
-  const docsCount =
+  const valueCount =
```
Fixed here [`80872e0` (#144646)](https://github.com/elastic/kibana/pull/144646/commits/80872e03e0de597573a005988aa03417e83cfb5a)

qn895 commented:
> Currently the counts in the expanded rows are not consistent:

@peteharverson Thanks for catching that. The numbers were adding up to the sampled # of docs for the numerical and boolean top stats, but not for string/keyword top terms. This is because the buckets' doc_count numbers returning from elasticsearch itself was not for the values sampled. I've added a fix for that by scaling the numbers down to match the # sampled docs (so `bucket.doc_count * (# sampled docs/# total docs)`. After the fix:
<img width="984" alt="Screen Shot 2022-11-14 at 17 41 04" src="https://user-images.githubusercontent.com/43350163/201791491-0a9b859b-ac6b-4675-819b-3c6b36d63fde.png">

> It would be good to add a debounce or delay to the sampling percentage slider.

@jgowdyelastic Good point! I've added debouncing here [`33c645f` (#144646)](https://github.com/elastic/kibana/pull/144646/commits/33c645fea169b20f2c47b022563a84eed816b2a3)

jgowdyelastic commented on x-pack/plugins/data_visualizer/public/application/common/components/document_count_content/document_count_content.tsx:
```diff
@@ -210,9 +219,7 @@ export const DocumentCountContent: FC<Props> = ({
-                        if (setSamplingProbability) {
-                          setSamplingProbability(closestProbability / 100);
-                        }
+                        updateSamplingProbability(closestProbability / 100);
```
it would be more performant to put all of the logic in this `onChange` function inside the function that is wrapped in the debounce. There's no need to be calculating the `closestProbability` on every change if it's just going to be discarded.

Also there is a useful hook called `useDebounce` which might work well here.
You could put the `e.currentTarget.value` in a temporary state variable e.g. `newProbability` and then `useDebounce` could watch for changes in that var.

peteharverson commented on x-pack/plugins/data_visualizer/public/application/common/components/top_values/top_values.tsx:
```diff
@@ -47,76 +46,54 @@ function getPercentLabel(docCount: number, topValuesSampleSize: number): string
-          id="xpack.dataVisualizer.dataGrid.field.topValues.calculatedFromSampleDescription"
-          defaultMessage="Calculated from sample of {topValuesSamplerShardSize} documents per shard"
+          id="xpack.dataVisualizer.dataGrid.field.topValues.calculatedFromSampleRecordsLabel"
+          defaultMessage="Calculated from {sampledDocumentsFormatted} sample {sampledDocuments, plural, one {record} other {records}}."
```
In the file data viz expanded rows, the doc count here always seems to be `0`:

<img width="789" alt="image" src="https://user-images.githubusercontent.com/7405507/201964267-09458f2c-243c-4a24-961a-8a6e15a78f77.png">

qn895 replied:
Fixed here [`7d7ba42` (#144646)](https://github.com/elastic/kibana/pull/144646/commits/7d7ba42c7872758583de72b127d08a9ea731696d)

peteharverson reviewed (approved):
Tested latest changes and LGTM.
Switch of position of the random sampler controls cog icon to the right looks good - this is consistent now with the position of the settings control in Discover and prevents the control jumping around as the slider is moved and the chart is reloaded.

jgowdyelastic reviewed (approved):
LGTM

qn895 commented:
@elasticmachine merge upstream

benwtrent commented on x-pack/plugins/data_visualizer/public/application/index_data_visualizer/search_strategy/requests/overall_stats.ts:
```diff
@@ -123,12 +123,17 @@ export const processAggregatableFieldsExistResponse = (
+      // https://www.elastic.co/guide/en/elasticsearch/reference/current/mapping-doc-count-field.html
+      // therefore we need to correct it by multiplying by the sampled probability
       const count = get(aggregations, [...aggsPath, `${safeFieldName}_count`, 'doc_count'], 0);
+      const multiplier =
+        count > sampleCount ? get(aggregations, [...aggsPath, 'probability'], 1) : 1;
```
same concerns here as above.

benwtrent commented on x-pack/plugins/data_visualizer/public/application/index_data_visualizer/search_strategy/requests/get_boolean_field_stats.ts:
```diff
@@ -80,15 +80,34 @@ export const fetchBooleanFieldsStats = (
 
         const batchStats: BooleanFieldStats[] = fields.map((field, i) => {
           const safeFieldName = field.fieldName;
+          // Sampler agg will yield doc_count that's bigger than the actual # of sampled records
+          // because it uses the stored _doc_count if available
+          // https://www.elastic.co/guide/en/elasticsearch/reference/current/mapping-doc-count-field.html
+          // therefore we need to correct it by multiplying by the sampled probability
```
I am not sure this is doing what you are commenting.

I think this is scaling the count back DOWN via the probability (random_sampler not only uses `_doc_count`, but scales sampled counting statistics by the inverted probability). In doing this, you ensure that doc count is near the sampledCount, but this has nothing to do with _doc_count.

benwtrent commented on x-pack/plugins/data_visualizer/public/application/index_data_visualizer/search_strategy/requests/utils.ts:
```diff
@@ -0,0 +1,42 @@
+  const topValues = topValuesBuckets.map((bucket) => ({
+    ...bucket,
+    doc_count: sampledCount
+      ? Math.floor(bucket.doc_count * (sampledCount / realNumberOfDocuments))
```
It would be good to have a comment here explaining `sampledCount / realNumberOfDocuments` that this was done to "scale back down" doc_count to be lower than the sampled number of documents. 

The sum of the bucket values is indeed to handle the `_doc_count` weirdness.
